### PR TITLE
Put the challenge at the end

### LINKED ISSUE #265: Andrew's comment
Tracking Issue for @andrewwhitehead's comment in PR #257: 

Comment: https://github.com/decentralized-identity/bbs-signature/pull/257#pullrequestreview-1468296014

### PR DESCRIPTION
**Pull request by BasileiosKal**
Fix #265

Move the challenge value `c` at the end of the proof value. Specifically, the proof becomes

```
proof = (Abar, Bbar, r2^, r3^, (m^_j1, ..., m^_jU), c)
```

Also change `Abar` with `Bbar` in ProofGen and `ProofVerify`

### CODE REVIEW

**Review by andrewwhitehead — APPROVED** [review]

**Comment by BasileiosKal** [comment]
## Breaking Changes

1. moved the challenge `c` value at the end of the proof value (requires an update to the proof serialization/de-serialization operations).
2. In `ProofGen`. At step 11 during `C` calculation, switched `Abar` with `Bbar`. I.e., 
    ```
    C = Bbar * r2 + Abar * r3 + ...  -->>  C = Abar * r2 + Bbar * r3 + ...
    ```
    Also in step 15 and 16 of `ProofGen`, the following changed (`e` "moved" from `r3^` to `r2^`)
    ```
    15. r2^ = r2 + r4 * c      -->>  r2^ = r2 + r4 * e * c
    16. r3^ = r3 + r4 * e * c  -->>  r3^ = r3 + r4 * c
    ```
3. In `ProofVerify`, at step 8, again change `Abar` with `Bbar` during `C` calculation, i.e., 
    ```
     C = Bbar * r2^ + Abar * r3^ + ...   -->>   C = Abar * r2^ + Bbar * r3^ + ...
    ```

**Comment by BasileiosKal** [comment]
@andrewwhitehead, I'm really sorry, but there are some more changes to review 🙏

**Comment by christianpaquin** [comment]
Is this captured in any test vectors?

**Review by christianpaquin — APPROVED** [review]
Some minor editorial comments. Implemented the update in my library ([PR 23](https://github.com/microsoft/bbs-node-reference/pull/23)); test vectors pass.

File: draft-irtf-cfrg-bbs-signatures.md [hunk 1]
```diff
@@ -1072,8 +1072,8 @@ Outputs:
 
 Procedure:
 
-1. (Abar, Bbar, c, r2^, r3^, (m^_1, ..., m^_U)) = proof
-2. return serialize((Abar, Bbar, c, r2^, r3^, m^_1, ..., m^_U))
+1. (Abar, Bbar, r2^, r3^, (m^_1, ..., m^_U), c) = proof
+2. return serialize((Abar, Bbar, r2^, r3^, m^_1, ..., m^_U), c)
```
[the remark this answers is not in the document]
Comment (BasileiosKal) [reply draft-irtf-cfrg-bbs-signatures.md]: Good catch! It was a typo. Fixed. Thanks!

File: draft-irtf-cfrg-bbs-signatures.md [hunk 2]
```diff
@@ -1122,7 +1122,7 @@ Procedure:
 7.      if A_i is INVALID or Identity_G1, return INVALID
 8.      index += octet_point_length
 
-// Scalars (i.e., (c, r2^, r3^, (m^_j1, ..., m^_jU)) in
+// Scalars (i.e., (r2^, r3^, (m^_j1, ..., m^_jU), c) in
```
[the remark this answers is not in the document]
Comment (BasileiosKal) [reply draft-irtf-cfrg-bbs-signatures.md]: Fixed. Thanks!

**Review by Wind4Greg — COMMENTED** [review]
I'm not sure why were are doing the swapping of Abar and Bbar. When I try to map this back to the [Revisiting BBS paper](https://eprint.iacr.org/2023/275.pdf) it seems like the way we currently have it is a close match and this change seems less like the paper. 

Is there another version of the paper with this changed? 

Is there a reason for this change?

Note that we also don't seem to cite the "Revisiting BBS" paper in the update.

**Comment by BasileiosKal** [comment]
Hey Greg 👋  The swapping is mainly for readability. Note that we just want to calculate `Abar * a_random_number + Bbar * another_random_number`. It is just a bit more sense to have `Abar * r2 + Bbar * r3` instead of `Bbar * r2 + Abar * r3`.

This does not go against the paper. It is just a naming convention. Another way to think of it, is instead of "swapping" `Abar` and `Bbar` we "renaming" `r2` to `r3`.

The reason it is a breaking change is that we have deterministic random number generation for testing.

**Review by Wind4Greg — APPROVED** [review]
Implemented and confirmed all test vectors. Note the high level text description still has the parameters in the old order:

```
The inputted proof value must consist of the following components, in that order:

1. Two (2) valid points of the G1 subgroup, different from the identity point of G1 (i.e., `Abar, Bbar`, in ProofGen)
2. Three (3) integers representing scalars in the range of 1 to r-1 inclusive (i.e., `c, r2^, r3^`, in ProofGen).
3. A number of integers representing scalars in the range of 1 to r-1 inclusive, corresponding to the undisclosed from the proof messages (i.e., `m^_j1, ..., m^_jU`, in ProofGen, where U the number of undisclosed messages).
```

**Comment by BasileiosKal** [comment]
Discussed on the WG call on the 10th of July. Test vectors are cross validated by multiple implementations. Merging.

**Pull request merged**